src/dgemm_kernel.rs

```diff
@@ -102,6 +110,442 @@ pub unsafe fn kernel_target_sse2(k: usize, alpha: T, a: *const T, b: *const T,
+    // ab[1] = a0 b1 | a1 b0 | a2 b3 | a3 b2
+    let a0b0_a1b0_a2b2_a3b2 = _mm256_blend_pd(ab[0], ab[1], 0b1010);
+    // Scheme a), step 0.1
+    let a0b1_a1b1_a2b3_a3b3 = _mm256_blend_pd(ab[1], ab[0], 0b1010);
+
+    // Scheme a), steps 0.2
+    // ab[2] = a0 b2 | a1 b3 | a2 b0 | a3 b1
+    // ab[3] = a0 b3 | a1 b2 | a2 b1 | a3 b0
+    let a0b2_a1b2_a2b0_a3b0 = _mm256_blend_pd(ab[2], ab[3], 0b1010);
+    // Scheme a), steps 0.3
+    let a0b3_a1b3_a2b1_a3b1 = _mm256_blend_pd(ab[3], ab[2], 0b1010);
+
+    // ab[4] = a4 b0 | a5 b1 | a6 b2 | a7 b3
+    // ab[5] = a4 b1 | a5 b0 | a6 b3 | a7 b2
+    let a4b0_a5b0_a6b2_a7b2 = _mm256_blend_pd(ab[4], ab[5], 0b1010);
+    let a4b1_a5b1_a6b3_a7b3 = _mm256_blend_pd(ab[5], ab[4], 0b1010);
+
+    // ab[6] = a0 b2 | a1 b3 | a2 b0 | a3 b1
+    // ab[7] = a0 b3 | a1 b2 | a2 b1 | a3 b0
+    let a4b2_a5b2_a6b0_a7b0 = _mm256_blend_pd(ab[6], ab[7], 0b1010);
+    let a4b3_a5b3_a6b1_a7b1 = _mm256_blend_pd(ab[7], ab[6], 0b1010);
+
+    // Scheme a), step 1.0
+    let a0b0_a1b0_a2b0_a3b0 = _mm256_permute2f128_pd(
+        a0b0_a1b0_a2b2_a3b2,
+        a0b2_a1b2_a2b0_a3b0,
+        0x30
+    );
+    // Scheme a), step 1.1
+    let a0b2_a1b2_a2b2_a3b2 = _mm256_permute2f128_pd(
+        a0b0_a1b0_a2b2_a3b2,
+        a0b2_a1b2_a2b0_a3b0,
+        0x12,
+    );
+    // Scheme a) step 1.2
+    let a0b1_a1b1_a2b1_a3b1 = _mm256_permute2f128_pd(
+        a0b1_a1b1_a2b3_a3b3,
+        a0b3_a1b3_a2b1_a3b1,
+        0x30
+    );
+    // Scheme a) step 1.3
+    let a0b3_a1b3_a2b3_a3b3 = _mm256_permute2f128_pd(
+        a0b1_a1b1_a2b3_a3b3,
+        a0b3_a1b3_a2b1_a3b1,
+        0x12
+    );
+
+    // As above, but for ab[4..7]
+    let a4b0_a5b0_a6b0_a7b0 = _mm256_permute2f128_pd(
+        a4b0_a5b0_a6b2_a7b2,
+        a4b2_a5b2_a6b0_a7b0,
+        0x30
+    );
+    let a4b2_a5b2_a6b2_a7b2 = _mm256_permute2f128_pd(
+        a4b0_a5b0_a6b2_a7b2,
+        a4b2_a5b2_a6b0_a7b0,
+        0x12,
+    );
+    let a4b1_a5b1_a6b1_a7b1 = _mm256_permute2f128_pd(
+        a4b1_a5b1_a6b3_a7b3,
+        a4b3_a5b3_a6b1_a7b1,
+        0x30
+    );
+    let a4b3_a5b3_a6b3_a7b3 = _mm256_permute2f128_pd(
+        a4b1_a5b1_a6b3_a7b3,
+        a4b3_a5b3_a6b1_a7b1,
+        0x12
+    );
+
+    ab[0] = a0b0_a1b0_a2b0_a3b0;
+    ab[1] = a0b1_a1b1_a2b1_a3b1;
+    ab[2] = a0b2_a1b2_a2b2_a3b2;
+    ab[3] = a0b3_a1b3_a2b3_a3b3;
+
+    ab[4] = a4b0_a5b0_a6b0_a7b0;
+    ab[5] = a4b1_a5b1_a6b1_a7b1;
+    ab[6] = a4b2_a5b2_a6b2_a7b2;
+    ab[7] = a4b3_a5b3_a6b3_a7b3;
+
+    // Compute α (A B)
+    // _mm256_set1_pd and _mm256_broadcast_sd seem to achieve the same thing.
+    let alpha_v = _mm256_broadcast_sd(&alpha);
+    loop_m!(i, ab[i] = _mm256_mul_pd(alpha_v, ab[i]));
+
+    macro_rules! c {
+        ($i:expr, $j:expr) =>
+            (c.offset(rsc * $i as isize + csc * $j as isize));
+    }
+
+    // C ← α A B + β C
+    // XXX: Explictly setting c might not be necessary if we don't handle
+    // the β = 0 case separately.
```

Comment:
I think please do handle it separately, otherwise we have the indirection in the masked kernel for every block, seems suboptimal.